test/e2e/framework/framework.go

```diff
@@ -42,28 +45,51 @@ var _ FullClientInterface = &Framework{}
 var _ ClusterInterface = &Framework{}
 
 func NewFramework(namePrefix string) *Framework {
+	var err error
+
 	f := &Framework{
-		name:       names.SimpleNameGenerator.GenerateName(fmt.Sprintf("%s-", namePrefix)),
-		FullClient: FullClient{},
+		name:            names.SimpleNameGenerator.GenerateName(fmt.Sprintf("%s-", namePrefix)),
+		e2eArtifactsDir: "",
+		FullClient:      FullClient{},
+		clusters:        make([]*Cluster, 0, len(TestContext.RestConfigs)),
 	}
 
-	clusters := make([]*Cluster, 0, len(TestContext.RestConfigs))
+	if len(TestContext.ArtifactsDir) != 0 {
+		f.e2eArtifactsDir = path.Join(TestContext.ArtifactsDir, "e2e")
+		err = os.Mkdir(f.e2eArtifactsDir, 0777)
+		if !os.IsExist(err) {
+			o.Expect(err).NotTo(o.HaveOccurred())
+		}
+	}
+
+	o.Expect(TestContext.RestConfigs).NotTo(o.BeEmpty())
 	for i, restConfig := range TestContext.RestConfigs {
 		clusterName := fmt.Sprintf("%s-%d", f.name, i)
+
+		clusterE2EArtifactsDir := ""
+		if len(f.e2eArtifactsDir) != 0 {
+			clusterE2EArtifactsDir = path.Join(f.e2eArtifactsDir, fmt.Sprintf("cluster-%d", i))
+			err = os.Mkdir(clusterE2EArtifactsDir, 0777)
+			if !os.IsExist(err) {
+				o.Expect(err).NotTo(o.HaveOccurred())
+			}
+		}
+
 		c := NewCluster(
 			clusterName,
+			clusterE2EArtifactsDir,
 			restConfig,
 			func(ctx context.Context, adminClient kubernetes.Interface, adminClientConfig *restclient.Config) (*corev1.Namespace, Client) {
 				return CreateUserNamespace(ctx, clusterName, f.CommonLabels(), adminClient, adminClientConfig)
 			},
 		)
-		clusters = append(clusters, c)
+		f.clusters = append(f.clusters, c)
 	}
-	f.clusters = clusters
 
-	f.FullClient.AdminClient.Config = f.defaultCluster().AdminClientConfig()
+	f.FullClient.AdminClient.Config = f.GetDefaultCluster().AdminClientConfig()
 
 	g.BeforeEach(f.beforeEach)
+	g.JustAfterEach(f.justAfterEach)
```

Comment:
I don't think we should further couple the framework initialisation with the ginkgo nodes as any of these prevents you from ensuring an order in the tests itself. Is there any benefit to doing this in JustAfterEach instead of the first action in AfterEach?

Reply:
you need an order in which everything is first collected and then separately cleaned up with independent failure domains.

```
/*
JustAfterEach nodes are similar to AfterEach nodes, however they are guaranteed to run *before* all AfterEach node closures - just after the It node closure. This can allow you to separate diagnostics collection from teardown for a spec.
```

Reply:
Ok, from correctness perspective it makes sense. I guess it won't affect the amount of effort required to decouple those much either.